### DISCUSSION TITLE
[Question] Best way to parse the json file to get a summary table of BGCs

### DISCUSSION THREAD

**Discussion by jolespin** [antismash/antismash#484] [Q&A]
For example, on the antismash webserver there are summary tables:

<img width="683" alt="image" src="https://user-images.githubusercontent.com/9061708/198853585-87f16b9f-973a-45e4-b0eb-b3fd72a275c8.png">

https://antismash-db.secondarymetabolites.org/output/GCA_000002515.1/

I was working on a parser but it got full of "if else then" statements to get this type of table.

Is there anything you recommend to get the data into tabular format even if it doesn't have all of the information in the json?  For example, I don't need to full sequences but knowing what types of BGCs are present on which contigs w/ the number of genes would be helpful.

**Comment by SJShaw** [comment]
For newer releases, you can get all the information in that image from the JSON in `records[].areas`. Each area has the region start and end, a list of products/cluster types as shown, and information on protoclusters and candidate clusters.

The JSON doesn't store number of genes in each region, since that doesn't generally mean anything. Nor does it store a flat list of cluster types for each record, since that would lose the context of what else they were found with and so on.

If you want number of genes, then I'd suggest using the genbank outputs with the antiSMASH source itself (only the python parts are necessary, the binaries wouldn't be used here) and doing something like [this previous suggestion](https://github.com/antismash/antismash/issues/128#issuecomment-480355708).

**Reply by jolespin** [reply]
Thanks for getting back so quickly. Yea, the number of genes isn't essential but mostly the BGC type, contig position, and whether it's on an edge (which I could figure out easily from the contig positions). 

Are there any scripts that are used or have been used in previous versions that parses the json file and instead of the gbk? I've been using the built in json parser to turn into a Python dictionary but it's very deep and found it difficult to make a parser that worked in all cases.  For example, it would work on a few BGCs for one genome but wouldn't work for BGCs from another genome.

**Comment by SJShaw** [comment]
For older versions of results, using the matching version of antiSMASH and `antismash.common.serialiser.AntismashResults` should parse things correctly. This is how the antiSMASH-database imported the data used in the database. The script for importing it is [publicly available](https://github.com/antismash/db-import/blob/22d97c9f714c4e379bddce4138544200471bbfa0/import_json.py#L77). If you aren't using results of the same version as that importer was set up for, 5.2.0, then the process might need to be tweaked quite a bit.

I'm not sure if this is what you meant by turning it into a python dictionary, but if you just wanted a dictionary to work with in python that's very simple with the standard `json` library.

**Comment by jolespin** [comment]
I coded this up in case it's useful for anyone: 
https://github.com/jolespin/veba/blob/main/src/scripts/antismash_genbanks_to_table.py

The output looks like this and it's indexed by genome, contig, region, and gene:

<img width="1691" alt="image" src="https://user-images.githubusercontent.com/9061708/199301492-9d5df48b-8ef0-484e-b7f4-336d199961f7.png">

I ended up just using the genbank files instead of the json.  I couldn't figure out how to get this type of information from the json file.  I loaded it using the `json` python module but going through the features, there was a lot of extra information and difficult to figure out how to pull out the BGC details.  Though, the genbank files had everything I needed.  The table format makes it a little easier to combine the results from multiple genomes in a metagenome, get some basic stats on how many BGCs, and also filter out the BGCs on contig edges.  Might not be very scalable to working with extremely large datasets.

**Comment by jolespin** [comment]
Is there a way to get the type of table above (doesn't have to be exact) but from the json instead of the .gbk files?
